Mathlib/MeasureTheory/Group/FundamentalDomain.lean

```diff
@@ -719,4 +727,315 @@ protected theorem fundamentalInterior : IsFundamentalDomain G (fundamentalInteri
+  simp [iUnion_inter]
+
+variable {G}
+
+@[to_additive NullMeasurableSet.addQuotientMeasure_apply]
+lemma NullMeasurableSet.quotientMeasure_apply {U : Set (Quotient α_mod_G)}
+    (meas_U : MeasurableSet U) : hs.quotientMeasure G μ U = μ ((π ⁻¹' U) ∩ s) :=
+  MeasureTheory.Measure.ofMeasurable_apply U meas_U
+
+@[to_additive NullMeasurableSet.addQuotientMeasure_eq_map_restrict]
+lemma NullMeasurableSet.quotientMeasure_eq_map_restrict :
+    hs.quotientMeasure G μ = Measure.map π (μ.restrict s) := by
+  ext U meas_U
+  rw [Measure.map_apply (f := π) (fun V hV ↦ measurableSet_quotient.mp hV) meas_U,
+    Measure.restrict_apply (t := (Quotient.mk α_mod_G ⁻¹' U)) (measurableSet_quotient.mp meas_U)]
+  exact hs.quotientMeasure_apply μ meas_U
+
+@[to_additive NullMeasurableSet.eq_addQuotientMeasure]
+lemma NullMeasurableSet.eq_quotientMeasure {ν : Measure (Quotient α_mod_G)}
+    (h : ∀ (U : Set (Quotient α_mod_G)) (_ : MeasurableSet U), ν U = μ (π ⁻¹' U ∩ s)) :
+    ν = hs.quotientMeasure G μ := by
+  ext V meas_V
+  rw [h V meas_V, hs.quotientMeasure_apply μ meas_V]
+
+@[to_additive IsAddFundamentalDomain.addQuotientMeasure_invariant]
+lemma IsFundamentalDomain.quotientMeasure_invariant [Countable G] [MeasurableSpace G] {t : Set α}
+    [SMulInvariantMeasure G α μ] [MeasurableSMul G α] (fund_dom_s : IsFundamentalDomain G s μ)
+    (fund_dom_t : IsFundamentalDomain G t μ) :
+    (fund_dom_s.nullMeasurableSet.quotientMeasure G μ)
+    = fund_dom_t.nullMeasurableSet.quotientMeasure G μ := by
+  apply fund_dom_t.nullMeasurableSet.eq_quotientMeasure
+  intro U meas_U
+  rw [fund_dom_s.nullMeasurableSet.quotientMeasure_apply μ meas_U]
+  apply MeasureTheory.IsFundamentalDomain.measure_set_eq fund_dom_s fund_dom_t
+  -- weird elaboration order reasons why this `exact` can't be folded into the `apply` without
+  -- angering `aesop` later?
+  · exact measurableSet_quotient.mp meas_U
+  · intro g
+    ext x
+    have : Quotient.mk α_mod_G (g • x) = Quotient.mk α_mod_G x
+    · apply Quotient.sound
+      use g
+    aesop
+
+end FundamentalDomainMeasure
+
+/-! ## `HasFundamentalDomain` typeclass
+
+We define `HasFundamentalDomain` in order to be able to define the `covolume` of a quotient of `α`
+by a group `G`, which under reasonable conditions does not depend on the choice of fundamental
+domain. Even though any "sensible" action should have a fundamental domain, this is a rather
+delicate question which was recently addressed by Misha Kapovich: https://arxiv.org/abs/2301.05325
+
+TODO: Formalize the existence of a Dirichlet domain as in Kapovich's paper.
+
+-/
+
+section HasFundamentalDomain
+
+/-- "We say a quotient of `α` by `G` `HasFundamentalDomain` if there is a measurable set
+  `s` for which `IsFundamentalDomain G s` holds." -/
+class HasAddFundamentalDomain (G : Type*) (α : Type*) [Zero G] [VAdd G α] [MeasureSpace α] :
+    Prop :=
+  (has_add_fundamental_domain_characterization : ∃ (s : Set α), IsAddFundamentalDomain G s)
```

Comment:
A style cleanup:
```suggestion
class HasAddFundamentalDomain (G α : Type*) [Zero G] [VAdd G α] [MeasureSpace α] : Prop where
  has_add_fundamental_domain_characterization : ∃ s : Set α, IsAddFundamentalDomain G s
```
assuming that fits within 100 chars. Same below

Reply:
Done, thanks!